libwire/src/IpHeader.cpp

```diff
@@ -1,38 +1,47 @@
+#include "IpHeader.h"
+
 #include <iostream>
+#include <cstring>
 
-#include <string.h>
 #include <arpa/inet.h>
 
-#include "IpHeader.h"
-
 namespace Net {
-	IpHeader::IpHeader(unsigned char *buffer) {
-		ip_header = (struct iphdr*)buffer;
-		
-		struct sockaddr_in source,dest;
-		memset(&source, 0, sizeof(source));
-		source.sin_addr.s_addr = ip_header->saddr;
-
-		memset(&dest, 0, sizeof(dest));
-		dest.sin_addr.s_addr = ip_header->daddr;
-		
-		saddr = std::string(inet_ntoa(source.sin_addr));
-		daddr = std::string(inet_ntoa(dest.sin_addr));
+
+IpHeader::IpHeader(data_t const& buffer) {
+    _ipHeader = *(reinterpret_cast<const iphdr_t*>(buffer.data()));
+    struct sockaddr_storage source, dest;
+    struct sockaddr_in* ipv4ptr;
+
+    std::memset(&source, 0, sizeof(source));
+    std::memset(&dest, 0, sizeof(dest));
+
+    ipv4ptr = reinterpret_cast<struct sockaddr_in*>(&source);
+    ipv4ptr->sin_addr.s_addr = _ipHeader.saddr;
+    ipv4ptr = reinterpret_cast<struct sockaddr_in*>(&dest);
+    ipv4ptr->sin_addr.s_addr = _ipHeader.daddr;
+
+    try {
+        _saddr = reinterpret_cast<struct sockaddr*>(&source);
```

Comment:
assign _saddr wich is now a Net::NetAddr, this assign can throw so we try catch.
I documented when a method can throw, look in NetAddr.h